sdk-base/src/main/java/com/mapbox/maps/plugin/viewport/experimental/ViewportPlugin.kt

```diff
@@ -0,0 +1,154 @@
+
+  // States
+
+  /**
+   * Returns list of registered states.
+   */
+  val states: Set<ViewportState>
+
+  /**
+   * Add a viewport state to the viewport plugin, which could be reused later with the state id.
+   *
+   * @param state the view port state to be added.
+   */
+  fun addState(state: ViewportState)
+
+  /**
+   * Remove a viewport state from the viewport plugin, and it could not be reused later with the state id.
+
+   * @param state the view port state to be removed.
+   */
+  fun removeState(state: ViewportState)
+
+  // Transitions
+
+  /**
+   * DefaultViewportTransition with default options
+   */
+  var defaultTransition: ViewportTransition
+
+  /**
+   * Set the [ViewportTransition] for the transition from given [ViewportState] to target [ViewportState]
+   *
+   * @param transition The transition to be set.
+   * @param from The state before the transition.
+   * @param to The state after the transition.
+   */
+  fun setTransition(transition: ViewportTransition, from: ViewportState?, to: ViewportState)
+
+  /**
+   * Get the transition from the previous state and target state.
+   *
+   * @param from The state before the transition.
+   * @param to The state after the transition.
+   */
+  fun getTransition(from: ViewportState?, to: ViewportState): ViewportTransition?
+
+  /**
+   * Remove the transition between the previous state and target state.
+   *
+   * @param from The state before the transition.
+   * @param to The state after the transition.
+   */
+  fun removeTransition(from: ViewportState?, to: ViewportState)
+
+  // Observers
+
+  /**
+   * Adds [ViewportStatusObserver].
+   */
+  fun addStatusObserver(
+    viewportStatusObserver: ViewportStatusObserver
+  )
+
+  /**
+   * Removes [ViewportStatusObserver].
+   */
+  fun removeStatusObserver(
+    viewportStatusObserver: ViewportStatusObserver
+  )
+
+  // Convenient methods to create the in-stock [ViewportState] and [ViewportTransition].
+
+  /**
+   * Create a [FollowingViewportState] instance with provided [FollowingViewportStateOptions].
+   *
+   * @param options The desired [FollowingViewportStateOptions]
+   */
+  fun makeFollowingViewportState(options: FollowingViewportStateOptions): FollowingViewportState
+
+  /**
+   * Create an [OverviewViewportState] instance with provided [OverviewViewportStateOptions].
+   *
+   * @param options The desired [OverviewViewportStateOptions]
+   */
+  fun makeOverviewViewportState(options: OverviewViewportStateOptions): OverviewViewportState
+
+  /**
+   * Create a default [ViewportTransition] instance with provided [DefaultViewportTransitionOptions].
+   *
+   * @param options The desired [DefaultViewportTransitionOptions]
+   */
+  fun makeDefaultTransition(options: DefaultViewportTransitionOptions): ViewportTransition
+
+  /**
+   * Create a [ViewportTransition] instance that transition to the target [ViewportState] immediately.
+   */
+  fun makeImmediateViewportTransition(): ViewportTransition
```

Comment:
Interesting that you made the return type the interface instead of the implementation here. I wanted to do the same on iOS for testing purposes, but felt that since we made the options on these transitions mutable developers would want the type to be the actual implementation type so they'd have access.

I'm completely open to making options immutable though and returning the interface from these methods.

I notice that you returned concrete types for the two state factory methods above. Curious whether there's a good reason to use different approaches in those cases.

Reply:
> I notice that you returned concrete types for the two state factory methods above. Curious whether there's a good reason to use different approaches in those cases.

The two state factory methods also is returned as interfaces:
```
interface FollowingViewportState : ViewportState {
  /**
   * Describes the configuration options of the state.
   */
  var options: FollowingViewportStateOptions
}
```
I just included the options inside the interface, so that use can modify these options later on.

I guess we could do the same for the default transition, wdyt?

Reply:
Refactored DefaultViewportTransition to interface that contains the DefaultViewportTransitionOptions, the same way as FollowingViewportState and OverviewViewportState.

Reply:
From discussion: some concerns were raised about the general Android pattern in v10 where interfaces are defined in sdk-base. For v11, we should revisit this and consider only exposing public interfaces in situations where developers would be expected to make custom implementations to provide to the SDK.

For iOS, we'll stick with returning concrete types from these methods.